clients/da-vinci-client/src/main/java/com/linkedin/davinci/consumer/VeniceAfterImageConsumerImpl.java

```diff
@@ -166,47 +181,58 @@ protected CompletableFuture<Void> internalSeek(
     return super.internalSeek(partitions, targetTopic, seekAction);
   }
 
-  private class VersionSwapDetectionThread implements Runnable {
+  private class VersionSwapDataChangeListener implements StoreDataChangedListener {
     @Override
-    public void run() {
-      Set<Integer> partitions = new HashSet<>();
-      try {
-        // Check the current version of the server
-        int currentVersion = storeRepository.getStore(storeName).getCurrentVersion();
-
-        // Check the current ingested version
-        Set<PubSubTopicPartition> subscriptions = getTopicAssignment();
-        synchronized (subscriptions) {
-          if (subscriptions.isEmpty()) {
-            return;
-          }
-          int maxVersion = -1;
-          for (PubSubTopicPartition topicPartition: subscriptions) {
-            int version = Version.parseVersionFromVersionTopicName(topicPartition.getPubSubTopic().getName());
-            if (version >= maxVersion) {
-              maxVersion = version;
+    public void handleStoreChanged(Store store) {
+      synchronized (this) {
+        Set<Integer> partitions = new HashSet<>();
+        try {
+          // Check the current version of the server
+          int currentVersion = storeRepository.getStore(storeName).getCurrentVersion();
+
+          // Check the current ingested version
+          Set<PubSubTopicPartition> subscriptions = getTopicAssignment();
+          synchronized (subscriptions) {
```

Comment:
Can you explain me the scenario when we need to synchronized on `subscriptions` when we have synchronized on this? I'm worried this may cause some deadlock.

Reply:
it's kind of a long story, but getTopicAssignment returns a synchronized set that wraps the assignment list returned by the pubsub consumer.  The pubsub consumer assignment listing is not thread safe.  It does like an assignments.values() wrapped to a hashmap, but that values is backed by the actual map that's in the pubsub consumer.  We wrap it to the synchronized set to prevent thrown exceptions for concurrent access, though, rereading the implementation of synchronized set, it occurs to me that the synchronized blocks here actually don't really do anything since getAssignments returns a new instance every time.  So it's harmless, but it's also not helpful.  Might just prefer to remove it.